src/ddmd/expressionsem.d

```diff
@@ -8771,6 +8784,94 @@ Expression binSemanticProp(BinExp e, Scope* sc)
     return null;
 }
 
+/********************************************************************************
+ * Helper function to consolidate common code for `BinAssingExp`s.
+ * It rewrites expressions of the form, for example, `s.x += 1` to `s.x(s.x() += 1)`
+ * if `s` has a both a getter and setter `@property` named `x`.
+ * Params:
+ *      e = the expression to rewrite
+ *      sc = the semantic scope
+ * Returns:
+ *      the rewritten expression if the procedure succeeds, an `ErrorExp` if the
+ *      and error is encountered, or `null` if `e.e1` is not a `DotIdExp`.
+ */
+Expression binSemanticProp(BinAssignExp e, Scope* sc)
+{
+    if (Expression ex = binSemantic(e, sc))
+        return ex;
+    Expression e1x = resolveProperties(sc, e.e1);
+    Expression e2x = resolveProperties(sc, e.e2);
+    if (e1x.op == TOKerror)
+        return e1x;
+    if (e2x.op == TOKerror)
+        return e2x;
+
+    //********************************************************************
+    // Check for property assignment.
+    // https://issues.dlang.org/show_bug.cgi?id=8006
+    if (e1x.op == TOKcall)
+    {
+        CallExp ce = cast(CallExp)e1x;
+
+        // Consider s.x() += 1;
+        // We want to rewrite that expression to s.x(s.x() + 1);
+        if (ce.e1.type.ty == Tfunction && (cast(TypeFunction)ce.e1.type).isproperty)
```

Comment:
Actually, I think it's best to have this check as:
```
auto tf = (ce.e1.type.ty) == Tfunction ? cast(TypeFunction)ce.e1.type : null;
if (tf && tf.isproperty && !tf.isref)
```
So that `ref @property` functions are *not* rewritten.